### DISCUSSION TITLE
Support non-PKCE OAuth2 flow

### DISCUSSION THREAD

**Discussion by nicolaschapeau** [directus/directus#12174] [Feature Requests]
Hi,

I am having trouble implementing oauth2 with a provider named "NetExplorer".
Everything looks good to me, but when i am redirected to directus, I get the following error : "INVALID_TOKEN".

I decided to enable "trace" mode in the logging system and i get this error : "Invalid grant".
I already used this API on a custom made nodeJS server and oauth2 was working fine. 
(There was a 'Content-Type': 'application/x-www-form-urlencoded' header i don't know if it can change something)

I noticed that directus was also adding a parameter "access_type=offline" while not supported by Netexplorer. But except that nothing out of the ordinary.

I linked my env and the documentation.

env
![env (1)](https://user-images.githubusercontent.com/32775238/158230345-8a95c413-0c37-4c75-8487-f9284eb39de2.PNG)

authorize
![documentation-authorize](https://user-images.githubusercontent.com/32775238/158229979-0f110b0e-f9fb-4c07-b120-295aa2463030.PNG)
![documentation-authorize-2](https://user-images.githubusercontent.com/32775238/158229978-1b2b5662-fd29-478a-a360-aa39b5a3bdf6.PNG)

access
![documentation-access-url](https://user-images.githubusercontent.com/32775238/158230051-d5bc8ef6-a736-408f-834a-7edcaaca77a1.PNG)
![documentation-access-url-2](https://user-images.githubusercontent.com/32775238/158230046-3c47f4be-3373-43e4-9305-da89756bb93d.PNG)

### To Reproduce

Directus self hosted with docker

Created a redirect_uri in cooperation with NetExplorer (private api)

Try to use NetExplorer as a provider

### Errors Shown

Error log on "trace" mode

![logs_after_redirect (1)](https://user-images.githubusercontent.com/32775238/158229355-8a945b26-c906-4681-bb18-e5a273129ed3.PNG)

After redirect from netexplorer to directus

![redirect (1)](https://user-images.githubusercontent.com/32775238/158229477-eb53aef9-7c25-44ba-acbe-5b4c59fc21ac.PNG)

### What version of Directus are you using?

9.5.1

### What version of Node.js are you using?

16.13.2

### What database are you using?

postgre

### What browser are you using?

chrome

### What operating system are you using?

linux

### How are you deploying Directus?

docker

**Comment by aidenfoxx** [comment]
My best guess is the "refresh_token" is being rejected by the provider. If you could enable `LOG_STYLE="raw"` we'll be able to see the full stack trace of the issue?

EDIT:

I can actually see in the logs that the users "refresh_token" is being rejected by netexplorer. When an oauth provider rejects the "refresh_token" (by throwing the "invalid_grant" error) we try and get a new "refresh_token" by forcing the user to re-authenticate with `?prompt="consent"`. We can see this happen in the logs when Directus calls `/login?continue&prompt=true` right after `/login?continue`.

So for some reason when calling your oauth provider with `?prompt="consent"` we're not getting a new "refresh_token".

**Comment by nicolaschapeau** [comment]
Thanks for helping,


Edit: Getting logs...

I noticed the behavior you described. When i am redirected for a first time to directus i am redirected back to netexplorer and finally back to directus after accepting everything.

I never tried the code_challenge/code_challenge_method with them. I don't know if there is a way to disable that to have a more basic case?

I was thinking of redoing the oauth2 myself with a customized endpoint if there is no solution.

**Comment by aidenfoxx** [comment]
Really you shouldn't need to use a custom implementation. The issue is that the "refresh_token" is becoming invalidated at some point and netexplorer isn't sending us a new one. Directus will always update the "refresh_token" if the provider sends us one... we just need to figure out how to get a new "refresh_token" from the provider, or if the token they're sending us is even valid to begin with?

If you could query netexplorer about those points it might help us address this issue. Right now the problem does appear to be on their side.

**Comment by nicolaschapeau** [comment]
Formatted some logs 
[test.txt](https://github.com/directus/directus/files/8247865/test.txt)

I'll contact them tomorrow and see what i can get. I'll come back there when it's done !

Thanks.

**Comment by aidenfoxx** [comment]
Looking at the detailed logs I can see that it's the initial token request that's failing. So for some reason the token endpoint just throws a grant error when we're trying to get the initial "access_token". 🤔 

I notice in their docs it says the token endpoint expects the parameter "coded", but that's pretty non-standard. The library we use ("openid-client") sends the standard "code" parameter. Or maybe that's just google translate being weird? Otherwise we seem complaint.

**Comment by nicolaschapeau** [comment]
I can confirm it's google translate being weird. But i remember having this problem earlier with their API.
I will do some testing ASAP and try to figure out in which case it returns this error on their side. + contact them when i know

**Comment by nicolaschapeau** [comment]
Hi,

I just simulated the directus integration with a custom made server in node.js.
Seems like NetExplorer doesn't support well code_challenge/code_challenge_method (PKCE) which is used by default by directus.
If I only use de client_secret it works well.
Is it possible to disable this behavior ? I will contact NetExplorer and see how we can fix this, but it could take some time.

**Comment by aidenfoxx** [comment]
Unfortunately we don't currently offer any other oauth flows. We use PKCE as it is the recommended flow. If you really need it you could modify the oauth provider in a local build until NetExplorer fix their implementation.

**Reply by nicolaschapeau** [reply]
Okay, thanks for your help ! Currently discussing with them so they can fix this.

**Reply by nicolaschapeau** [reply]
Quick feedback :

Their API got fixed so everything is working well now.

Thanks !

**Reply by rijkvanzanten** [reply]
Awesome! Glad to hear that 🙂

**Reply by xu4wang** [reply]
It' great they fix it so fast.

I was integrating with Dingtalk(oauth provider) which does not support PKCE too. I made a proxy between directus and the oauth provider,  and store the client_secret in the proxy, added it [on the fly](https://github.com/xu4wang/apiproxy/blob/main/src/handlers/oauth_dingtalk.ts) when Directus requests for the token.

**Comment by rijkvanzanten** [comment]
I'll move this to a discussion, as it's technically discussing a different/new oauth flow that's considered a new feature 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
